build-utils/src/index.ts

```diff
@@ -0,0 +1,149 @@
+        cwd: projectDir,
+        stdio: 'inherit',
+        shell,
+      });
+    }
+    await spawn('npm', ['run', 'build'], {
+      cwd: projectDir,
+      stdio: 'inherit',
+      shell,
+    });
+    logger.log('Preparing backend...');
+    const deps = await getDependencies(projectDir) as Map<string, {optional: boolean}>;
+    for (const [dep, props] of deps) {
+      const source = pathResolve(projectDir, 'node_modules', dep);
+      const target = pathResolve(stagingDir, 'node_modules', dep);
+      if (await exists(source)) {
+        await mkdirp(target);
+        await copy(source, target);
+      } else if (!props.optional) {
+        logger.error(`WARN: Cannot find dependency ${dep} in node_modules, skipping upload`);
+      }
+    }
+
+    const filesToExclude = new Set(['node_modules', 'backend', 'src'].map((f) => pathResolve(projectDir, f)));
+    await copy(projectDir, stagingDir, {
+      filter(src) {
+        return !filesToExclude.has(src) &&
+          !(path.dirname(src) === projectDir && path.basename(src).startsWith('.'));
+      },
+    });
+
+    await spawn(escapeWin32(pathResolve(projectDir, 'node_modules', '.bin', 'babel')), [
+      '--no-babelrc',
+      '--config-file',
+      require.resolve('./babelBackendConfig.js'),
+      '--source-maps',
+      'true',
+      '--plugins',
+      ['@babel/plugin-transform-modules-commonjs',
+        'module:@reshuffle/code-transform'].join(','),
+      'backend/',
+      '-d',
+      escapeWin32(pathResolve(stagingDir, 'backend')),
+    ], {
+      cwd: projectDir,
+      stdio: 'inherit',
+      shell,
+    });
+
+    await copy(pathResolve(projectDir, 'backend'), pathResolve(stagingDir, 'backend'), {
+      filter(src) {
+        return path.extname(src) !== '.js';
+      },
+    });
+
+    return stagingDir;
+
+  } catch (err) {
+    await remove(stagingDir);
+    throw err;
+  }
+
+}
+
+export async function createTarball(stagingDir: string): Promise<string> {
+  const tarPath = pathResolve(stagingDir, 'bundle.tgz');
+  await tar.create({
+    gzip: true,
+    file: tarPath,
+    cwd: stagingDir,
+    filter: (filePath) => filePath !== './bundle.tgz',
+  }, ['.']);
+  return tarPath;
+}
+
+export async function uploadCode(
+  tarPath: string,
+  url: string,
+  apiHeaders: Record<string, string> = {},
+  logger = DEFAULT_OPTIONS.logger
+) {
+  const { size: contentLength } = await stat(tarPath);
+  const stream = createReadStream(tarPath);
+  logger.log('Uploading your assets! This may take a few moments, please wait');
+  const res = await fetch(url, {
+    method: 'POST',
+    headers: {
+      ...apiHeaders,
+      'Content-Type': 'application/gzip',
+      'Content-Length': `${contentLength}`,
+    },
+    body: stream,
+  });
+  const json = await res.json();
+  if (res.status !== 200) {
+    // TODO: check error if response is not a json and display a nice error message
+    const { message } = await res.json();
+    logger.error(message);
```

Comment:
Due to comment in L145 we might never reach this line, but when we do, previous behavior was `oclif.error` which logs and then throws a `CLIError` and just logs.

Reply:
I've handled this scenario but catching the error and re-throwing the `CLIError` in the `cli` project
https://github.com/reshufflehq/reshuffle/pull/377/files#diff-e88bfc89b1330c5aed4c81329ea15882R99

Reply:
I think this is a different kind of error - `logger.error` can be replaced with a ```throw new Error(`failed to upload: ${message}`);``` or similar.

Example of error response status is 500 and body is `{"message": "entity.too.large"}`.